extensions/src/main/java/org/pentaho/platform/web/http/api/resources/ActionResource.java

```diff
@@ -66,22 +65,44 @@
    * @return a {@link Response}
    */
   @POST
-  @Path( "/runInBackground" )
+  @Path( "/invoke" )
   @Consumes( { TEXT_PLAIN } )
   @StatusCodes( {
       @ResponseCode( code = 200, condition = "Action invoked successfully." ),
       @ResponseCode( code = 400, condition = "Bad input - could not invoke action." ),
       @ResponseCode( code = 401, condition = "User does not have permissions to invoke action" ),
       @ResponseCode( code = 500, condition = "Error while retrieving system resources." ),
     } )
-  public Response runInBackground(
+  public Response invokeAction(
+    @QueryParam( ActionUtil.INVOKER_ASYNC_EXEC ) @DefaultValue( ActionUtil.INVOKER_DEFAULT_ASYNC_EXEC_VALUE ) String async,
     @QueryParam( ActionUtil.INVOKER_ACTIONID ) String actionId,
     @QueryParam( ActionUtil.INVOKER_ACTIONCLASS ) String actionClass,
     @QueryParam( ActionUtil.INVOKER_ACTIONUSER ) String actionUser,
     final String actionParams ) {
 
-    executorService.submit( createRunnable( actionId, actionClass, actionUser, actionParams ) );
-    return Response.status( HttpStatus.SC_ACCEPTED ).build();
+    // https://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html#parseBoolean(java.lang.String)
+    boolean isAsyncExecution = Boolean.parseBoolean( async );
+    int httpStatus = HttpStatus.SC_INTERNAL_SERVER_ERROR; // default ( pessimistic )
+
+    if ( isAsyncExecution ) {
+
+      // default scenario for execution
+      executorService.submit( createRunnable( actionId, actionClass, actionUser, actionParams ) );
+      httpStatus = HttpStatus.SC_ACCEPTED;
+
+    } else {
+
+      try {
+
+        IActionInvokeStatus status = internalInvokeAction( actionId, actionClass, actionUser, actionParams );
```

Comment:
If I understand correctly, the purpose of introducing the internalInvokeAction() methods is to be able to get back the status of the run, that we don't have in Runnable. Instead, why not consider refactoring createRunnable to be a createCallable that returns a Callable.

Seems cleaner, and the asych vs sync will just invoke directly on the callable...like so:

    if (isAsyncExecution) {
    .... submit( createCallable(...) );
    }
.... else {
    ... IActionInvokeStatus status = createCallable(...).call();
}